nixos/modules/services/web-apps/jitsi-meet.nix

```diff
@@ -0,0 +1,503 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.jitsi-meet;
+  attrsToArgs = a: concatStringsSep " " (mapAttrsToList (k: v: "${k}=${toString v}") a);
+
+  # The configuration files are JS of format "var <<string>> = <<JSON>>;". In order to
+  # override only some settings, we need to extract the JSON, use jq to merge it with
+  # the config provided by user, and then reconstruct the file.
+  overrideJs =
+    source: varName: userCfg:
+    let
+      extractor = pkgs.writeText "extractor.js" ''
+        var fs = require("fs");
+        eval(fs.readFileSync(process.argv[2], 'utf8'));
+        process.stdout.write(JSON.stringify(eval(process.argv[3])));
+      '';
+      userJson = pkgs.writeText "user.json" (builtins.toJSON userCfg);
+    in (pkgs.runCommand "${varName}.js" { } ''
+      ${pkgs.nodejs}/bin/node ${extractor} ${source} ${varName} > default.json
+      (
+        echo "var ${varName} = "
+        ${pkgs.jq}/bin/jq -s '.[0] * .[1]' default.json ${userJson}
+        echo ";"
+      ) > $out
+    '');
+
+  # Essential config - it's probably not good to have these as option default because
+  # types.attrs doesn't do merging. Let's merge explicitly, can still be overriden if
+  # user desires.
+  defaultCfg = {
+    hosts = {
+      domain = cfg.hostName;
+      muc = "conference.${cfg.hostName}";
+    };
+    bosh = "//${cfg.hostName}/http-bind";
+  };
+
+  # HOCON is a JSON superset that videobridge2 uses for configuration.
+  # It can substitute environment variables which we use for passwords here.
+  # https://github.com/lightbend/config/blob/master/README.md
+  toHOCON = v: builtins.replaceStrings [ ''"__ENV__'' ''__ENV__"'' ] [ "\${" "}" ] (builtins.toJSON v);
```

Comment:
why don't we just use the native `${SOME_VAR}` syntax when assembling the config, instead of adding an intermediate nix stage?

Reply:
I'm afraid I don't understand the proposed solution.

Do you mean using `${SOME_VAR}` instead of `__ENV__SOME_VAR__ENV__` to represent variables internally? We'd still have to have a step that removes the quotes, transforming `{ "foo": "${BAR}" }` to `{ "foo": ${BAR} }`, as HOCON doesn't interpolate environment variables inside string literals.

Or do you mean constructing the configuration directly from strings? Then it won't be possible for the user to customize it through `services.jitsi-meet.config`.

Can you elaborate a bit please?

Reply:
Aaah… so the problem here is that `builtins.toJSON` will quote all strings, and HOCON won't replace environment vars if they're inside double quotes?

Reply:
Exactly - it's almost unbelievable if you've been staring at nix expressions long enough:grinning:

I guess the proper solution would be to design an encoding of HOCON to nested attrsets and providing functions to construct such expression and serialize it. Not sure whether that's worth the effort if no other module uses it.

Perhaps less hacky solution without the need for assert would be to represent the environment variable by attrset such as `{ __hocon_envvar = "SOMEVARNAME"; }` and implement `toHOCON` by if-elseif-elseif... chain like `lib.generators.toPretty`?

Reply:
Whatever we do, this specific way how we escape env vars to later replace them needs an explanatory comment.

Reply:
Changed this to the attrset solution described above which doesn't require assert. Please let me know if the attached explanation is sufficient.